api/pkg/oauth/auth/service/auth.go

```diff
@@ -0,0 +1,193 @@
+		api:     api,
+	}
+}
+
+// Return name and status of the services
+func Status(res http.ResponseWriter, req *http.Request) {
+
+	authSvc := Services{
+		AuthService{
+			Name:   "auth",
+			Status: "ok",
+		},
+	}
+
+	var log log.Logger
+	res.WriteHeader(http.StatusOK)
+	if err := json.NewEncoder(res).Encode(authSvc); err != nil {
+		log.Error(err)
+	}
+}
+
+// Authenticates user with the speicified git provider
+// using goth and calls the AuthCallback function
+func Authenticate(res http.ResponseWriter, req *http.Request) {
+	UI_URL = req.FormValue("redirect_uri")
+	gothic.BeginAuthHandler(res, req)
+}
+
+// Once user is authenticated, store the user details in db
+// and redirect to UI with the status code and auth code
+func (s *service) AuthCallBack(res http.ResponseWriter, req *http.Request) {
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	ghUser, err := gothic.CompleteUserAuth(res, req)
+	if err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	params := req.URL.Query()
+
+	if err = r.insertData(ghUser, params.Get("code")); err != nil {
+		r.log.Error(err)
+		res.Header().Set("Location", fmt.Sprintf("%s?status=%d", UI_URL, http.StatusBadRequest))
+		res.WriteHeader(http.StatusTemporaryRedirect)
+	}
+
+	res.Header().Set("Location", fmt.Sprintf("%s?status=%d&code=%s", UI_URL, http.StatusOK, params.Get("code")))
+	res.WriteHeader(http.StatusTemporaryRedirect)
+}
+
+// Checks for auth code in the headers, validates the
+// auth code and returns the jwt token for user
+func (s *service) HubAuthenticate(res http.ResponseWriter, req *http.Request) {
+
+	// Get the auth code from params
+	code := req.FormValue("code")
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	var User model.User
+	// Check if user exist
+	q := r.db.Model(&model.User{}).
+		Where("code = ?", code)
+
+	err := q.First(&User).Error
+	if err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	if err := r.db.Model(&model.User{}).Where("github_login = ?", User.GithubLogin).Update("code", "").Error; err != nil {
+		r.log.Error(err)
+	}
+
+	// gets user scopes to add in jwt
+	scopes, err := r.userScopes(&User)
+	if err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	user, err := r.createTokens(&User, scopes)
+	if err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	res.WriteHeader(http.StatusOK)
+	if err := json.NewEncoder(res).Encode(user); err != nil {
+		r.log.Error(err)
+		http.Error(res, err.Error(), http.StatusBadRequest)
```

Comment:
why bad req if it fails to encode in json?

Reply:
How about internal server error ??